### PR TITLE
[LLVM][rtsan] Add `sanitize_realtime_unsafe` attribute

### PR DESCRIPTION
**Pull request by davidtrevelyan**
# Introducing `sanitize_realtime_unsafe` attribute

## Motivation

Calls to system library functions such as `malloc` are easy for RealtimeSanitizer to intercept. If such a call is made in a `[[clang::nonblocking]]` function (a real-time context), RealtimeSanitizer will error. Real-time programmers also write their own blocking (real-time unsafe) functions that may or may not call intercepted functions. We wish to introduce a mechanism whereby RealtimeSanitizer can error on calls to these, too, if called within a real-time context.

## Implementation

At the same time as introducing `[[clang::nonblocking]]`, the `[[clang::blocking]]` attribute was also introduced. With the performance constraints attribute warnings (as errors) activated, blocking functions cannot be called from non-blocking functions, and this is enforced at compile time. The purpose of this PR is to introduce similar functionality into RealtimeSanitizer, so that it can make the equivalent check at run time.

Using the same mechanism as adding the `sanitize_realtime` LLVM function attribute to the IR of `[[clang::nonblocking]]` functions, we wish to add the `sanitize_realtime_unsafe` attribute. If this attribute is detected, the RealtimeSanitizer pass inserts a call to `__rtsan_expect_not_realtime`.

## Integration Roadmap

The above functionality is currently split into two patches.

- [ ] Introduction of the `sanitize_realtime_unsafe` LLVM attribute (this PR), and
- [ ] An update to RealtimeSanitizer's LLVM pass to use it (will submit a follow-up PR in future)

### CODE REVIEW

**Comment by davidtrevelyan** [comment]
Tagging @cjappl for review, as well as interested parties from similar past reviews: @vitalybuka @MaskRay @dougsonos @nikic @pcc

**Review by cjappl — APPROVED** [review]
LGTM - please wait for another reviewer to sign off before merging

**Review by vitalybuka — APPROVED** [review]

**Comment by vitalybuka** [comment]
Leaving to @cjappl to land

**Review by MaskRay — APPROVED** [review]

**Pull request merged**
